local function InteractiveToast(
	HeaderText: string,
	ContentText: string,
	Button1Text: string,
	Button2Text: string
) : (ScreenGui, ImageButton, ImageButton)
	local ActiveToastObject = Instance.new("ScreenGui")
	ActiveToastObject.Parent = nil --[[ e.g. PlayerGui ]]
	ActiveToastObject.Enabled = true
	ActiveToastObject.SafeAreaCompatibility = Enum.SafeAreaCompatibility.FullscreenExtension
	ActiveToastObject.ResetOnSpawn = false
	ActiveToastObject.IgnoreGuiInset = false
	ActiveToastObject.Name = "InteractiveToast"
	ActiveToastObject.DisplayOrder = 0
	ActiveToastObject.ZIndexBehavior = Enum.ZIndexBehavior.Sibling
	ActiveToastObject.ScreenInsets = Enum.ScreenInsets.CoreUISafeInsets
	ActiveToastObject.ClipToDeviceSafeArea = true
	local Frame_2241 = Instance.new("Frame")
	Frame_2241.Parent = ActiveToastObject
	Frame_2241.LayoutOrder = 0
	Frame_2241.BorderMode = Enum.BorderMode.Outline
	Frame_2241.Interactable = true
	Frame_2241.AnchorPoint = Vector2.new(0, 0)
	Frame_2241.BorderColor = BrickColor.new("Black")
	Frame_2241.SizeConstraint = Enum.SizeConstraint.RelativeXY
	Frame_2241.ZIndex = 1
	Frame_2241.AutomaticSize = Enum.AutomaticSize.Y
	Frame_2241.Size = UDim2.new(1, 0, 1, 0)
	Frame_2241.Draggable = false
	Frame_2241.Visible = true
	Frame_2241.ClipsDescendants = false
	Frame_2241.BorderColor3 = Color3.fromRGB(27, 42, 53)
	Frame_2241.Selectable = false
	Frame_2241.Style = Enum.FrameStyle.Custom
	Frame_2241.SelectionOrder = 0
	Frame_2241.Rotation = 0
	Frame_2241.Transparency = 1
	Frame_2241.BackgroundTransparency = 1
	Frame_2241.Position = UDim2.new(0, 0, 0, 0)
	Frame_2241.Active = false
	Frame_2241.Name = "1"
	Frame_2241.BorderSizePixel = 1
	Frame_2241.BackgroundColor3 = Color3.fromRGB(163, 162, 165)
	local Child = Instance.new("Frame")
	Child.Parent = Frame_2241
	Child.LayoutOrder = 0
	Child.BorderMode = Enum.BorderMode.Outline
	Child.Interactable = true
	Child.AnchorPoint = Vector2.new(0.5, 0.5)
	Child.BorderColor = BrickColor.new("Black")
	Child.SizeConstraint = Enum.SizeConstraint.RelativeXY
	Child.ZIndex = 1
	Child.AutomaticSize = Enum.AutomaticSize.Y
	Child.Size = UDim2.new(0, 360, 0, 0)
	Child.Draggable = false
	Child.Visible = true
	Child.ClipsDescendants = false
	Child.BorderColor3 = Color3.fromRGB(27, 42, 53)
	Child.Selectable = false
	Child.Style = Enum.FrameStyle.Custom
	Child.SelectionOrder = 0
	Child.Rotation = 0
	Child.Transparency = 0.30000001192092896
	Child.BackgroundTransparency = 0.30000001192092896
	Child.Position = UDim2.new(0.5, 0, 0.5, 0)
	Child.Active = false
	Child.Name = "Child"
	Child.BorderSizePixel = 0
	Child.BackgroundColor3 = Color3.fromRGB(0, 0, 0)
	local ListLayout = Instance.new("UIListLayout")
	ListLayout.Parent = Child
	ListLayout.FillDirection = Enum.FillDirection.Horizontal
	ListLayout.HorizontalAlignment = Enum.HorizontalAlignment.Left
	ListLayout.HorizontalFlex = Enum.UIFlexAlignment.None
	ListLayout.VerticalAlignment = Enum.VerticalAlignment.Top
	ListLayout.SortOrder = Enum.SortOrder.LayoutOrder
	ListLayout.Name = "ListLayout"
	ListLayout.Wraps = false
	ListLayout.Padding = UDim.new(0, 8)
	ListLayout.ItemLineAlignment = Enum.ItemLineAlignment.Automatic
	ListLayout.VerticalFlex = Enum.UIFlexAlignment.None
	local Padding = Instance.new("UIPadding")
	Padding.Parent = Child
	Padding.PaddingTop = UDim.new(0, 12)
	Padding.Name = "Padding"
	Padding.PaddingBottom = UDim.new(0, 12)
	Padding.PaddingLeft = UDim.new(0, 12)
	Padding.PaddingRight = UDim.new(0, 12)
	local Content = Instance.new("Frame")
	Content.Parent = Child
	Content.LayoutOrder = 2
	Content.BorderMode = Enum.BorderMode.Outline
	Content.Interactable = true
	Content.AnchorPoint = Vector2.new(0, 0)
	Content.BorderColor = BrickColor.new("Black")
	Content.SizeConstraint = Enum.SizeConstraint.RelativeXY
	Content.ZIndex = 1
	Content.AutomaticSize = Enum.AutomaticSize.Y
	Content.Size = UDim2.new(0, 0, 0, 0)
	Content.Draggable = false
	Content.Visible = true
	Content.ClipsDescendants = false
	Content.BorderColor3 = Color3.fromRGB(27, 42, 53)
	Content.Selectable = false
	Content.Style = Enum.FrameStyle.Custom
	Content.SelectionOrder = 0
	Content.Rotation = 0
	Content.Transparency = 1
	Content.BackgroundTransparency = 1
	Content.Position = UDim2.new(0, 0, 0, 0)
	Content.Active = false
	Content.Name = "Content"
	Content.BorderSizePixel = 0
	Content.BackgroundColor3 = Color3.fromRGB(163, 162, 165)
	local FlexItem = Instance.new("UIFlexItem")
	FlexItem.Parent = Content
	FlexItem.Name = "FlexItem"
	FlexItem.FlexMode = Enum.UIFlexMode.Fill
	FlexItem.ShrinkRatio = 0
	FlexItem.GrowRatio = 0
	FlexItem.ItemLineAlignment = Enum.ItemLineAlignment.Center
	local ListLayout = Instance.new("UIListLayout")
	ListLayout.Parent = Content
	ListLayout.FillDirection = Enum.FillDirection.Vertical
	ListLayout.HorizontalAlignment = Enum.HorizontalAlignment.Left
	ListLayout.HorizontalFlex = Enum.UIFlexAlignment.None
	ListLayout.VerticalAlignment = Enum.VerticalAlignment.Top
	ListLayout.SortOrder = Enum.SortOrder.LayoutOrder
	ListLayout.Name = "ListLayout"
	ListLayout.Wraps = false
	ListLayout.Padding = UDim.new(0, 2)
	ListLayout.ItemLineAlignment = Enum.ItemLineAlignment.Automatic
	ListLayout.VerticalFlex = Enum.UIFlexAlignment.None
	local Actions = Instance.new("Frame")
	Actions.Parent = Content
	Actions.LayoutOrder = 3
	Actions.BorderMode = Enum.BorderMode.Outline
	Actions.Interactable = true
	Actions.AnchorPoint = Vector2.new(0, 0)
	Actions.BorderColor = BrickColor.new("Black")
	Actions.SizeConstraint = Enum.SizeConstraint.RelativeXY
	Actions.ZIndex = 1
	Actions.AutomaticSize = Enum.AutomaticSize.Y
	Actions.Size = UDim2.new(1, 0, 0, 0)
	Actions.Draggable = false
	Actions.Visible = true
	Actions.ClipsDescendants = false
	Actions.BorderColor3 = Color3.fromRGB(27, 42, 53)
	Actions.Selectable = false
	Actions.Style = Enum.FrameStyle.Custom
	Actions.SelectionOrder = 0
	Actions.Rotation = 0
	Actions.Transparency = 1
	Actions.BackgroundTransparency = 1
	Actions.Position = UDim2.new(0, 0, 0, 0)
	Actions.Active = false
	Actions.Name = "Actions"
	Actions.BorderSizePixel = 0
	Actions.BackgroundColor3 = Color3.fromRGB(163, 162, 165)
	local ListLayout = Instance.new("UIListLayout")
	ListLayout.Parent = Actions
	ListLayout.FillDirection = Enum.FillDirection.Horizontal
	ListLayout.HorizontalAlignment = Enum.HorizontalAlignment.Center
	ListLayout.HorizontalFlex = Enum.UIFlexAlignment.Fill
	ListLayout.VerticalAlignment = Enum.VerticalAlignment.Top
	ListLayout.SortOrder = Enum.SortOrder.LayoutOrder
	ListLayout.Name = "ListLayout"
	ListLayout.Wraps = false
	ListLayout.Padding = UDim.new(0, 8)
	ListLayout.ItemLineAlignment = Enum.ItemLineAlignment.Automatic
	ListLayout.VerticalFlex = Enum.UIFlexAlignment.None
	local ToastButton1 = Instance.new("ImageButton")
	ToastButton1.Parent = Actions
	ToastButton1.LayoutOrder = 1
	ToastButton1.Active = true
	ToastButton1.Interactable = true
	ToastButton1.SizeConstraint = Enum.SizeConstraint.RelativeXY
	ToastButton1.ZIndex = 1
	ToastButton1.BorderSizePixel = 0
	ToastButton1.SliceCenter = Rect.new(0, 0, 0, 0)
	ToastButton1.Draggable = false
	ToastButton1.ScaleType = Enum.ScaleType.Stretch
	ToastButton1.Modal = false
	ToastButton1.AutoButtonColor = false
	ToastButton1.Transparency = 1
	ToastButton1.SelectionOrder = 0
	ToastButton1.BackgroundColor3 = Color3.fromRGB(0, 0, 0)
	ToastButton1.ImageTransparency = 0
	ToastButton1.Selectable = true
	ToastButton1.AnchorPoint = Vector2.new(0, 0)
	ToastButton1.Image = ""
	ToastButton1.TileSize = UDim2.new(1, 0, 1, 0)
	ToastButton1.ImageRectSize = Vector2.new(0, 0)
	ToastButton1.Selected = false
	ToastButton1.AutomaticSize = Enum.AutomaticSize.X
	ToastButton1.Size = UDim2.new(1, 0, 0, 36)
	ToastButton1.Visible = true
	ToastButton1.HoverImage = ""
	ToastButton1.Style = Enum.ButtonStyle.Custom
	ToastButton1.ClipsDescendants = true
	ToastButton1.BorderColor3 = Color3.fromRGB(27, 42, 53)
	ToastButton1.BorderMode = Enum.BorderMode.Outline
	ToastButton1.ImageColor3 = Color3.fromRGB(255, 255, 255)
	ToastButton1.PressedImage = ""
	ToastButton1.Rotation = 0
	ToastButton1.ImageRectOffset = Vector2.new(0, 0)
	ToastButton1.BackgroundTransparency = 1
	ToastButton1.Position = UDim2.new(0, 0, 0, 0)
	ToastButton1.SliceScale = 1
	ToastButton1.ResampleMode = Enum.ResamplerMode.Default
	ToastButton1.BorderColor = BrickColor.new("Black")
	ToastButton1.Name = "ToastButton1"
	local Stroke = Instance.new("UIStroke")
	Stroke.Parent = ToastButton1
	Stroke.ApplyStrokeMode = Enum.ApplyStrokeMode.Contextual
	Stroke.LineJoinMode = Enum.LineJoinMode.Round
	Stroke.Name = "Stroke"
	Stroke.Color = Color3.fromRGB(255, 255, 255)
	Stroke.Transparency = 0
	Stroke.Thickness = 1
	Stroke.Enabled = true
	local Text = Instance.new("TextLabel")
	Text.Parent = ToastButton1
	Text.LayoutOrder = 2
	Text.FontSize = Enum.FontSize.Size24
	Text.TextDirection = Enum.TextDirection.Auto
	Text.Active = false
	Text.Interactable = true
	Text.TextStrokeTransparency = 1
	Text.TextTruncate = Enum.TextTruncate.AtEnd
	Text.SizeConstraint = Enum.SizeConstraint.RelativeXY
	Text.ZIndex = 1
	Text.BorderSizePixel = 0
	Text.Draggable = false
	Text.RichText = true
	Text.SelectionOrder = 0
	Text.TextYAlignment = Enum.TextYAlignment.Center
	Text.TextScaled = false
	Text.BackgroundColor3 = Color3.fromRGB(163, 162, 165)
	Text.FontFace = Font.new("rbxasset://fonts/families/BuilderSans.json", Enum.FontWeight.Bold, Enum.FontStyle.Normal)
	Text.BorderMode = Enum.BorderMode.Outline
	Text.Selectable = false
	Text.OpenTypeFeatures = ""
	Text.AnchorPoint = Vector2.new(0, 0)
	Text.BorderColor = BrickColor.new("Black")
	Text.Visible = true
	Text.TextXAlignment = Enum.TextXAlignment.Center
	Text.AutomaticSize = Enum.AutomaticSize.X
	Text.Size = UDim2.new(1, 0, 1, 0)
	Text.TextWrapped = true
	Text.TextWrap = true
	Text.TextTransparency = 0
	Text.ClipsDescendants = false
	Text.BorderColor3 = Color3.fromRGB(27, 42, 53)
	Text.Text = Button1Text
	Text.TextStrokeColor3 = Color3.fromRGB(0, 0, 0)
	Text.LineHeight = 1
	Text.Rotation = 0
	Text.Font = Enum.Font.BuilderSansBold
	Text.BackgroundTransparency = 1
	Text.Position = UDim2.new(0, 0, 0, 0)
	Text.TextSize = 20
	Text.MaxVisibleGraphemes = -1
	Text.TextColor3 = Color3.fromRGB(255, 255, 255)
	Text.Name = "Text"
	local CornerRadius = Instance.new("UICorner")
	CornerRadius.Parent = ToastButton1
	CornerRadius.Name = "CornerRadius"
	CornerRadius.CornerRadius = UDim.new(0, 8)
	local ListLayout = Instance.new("UIListLayout")
	ListLayout.Parent = ToastButton1
	ListLayout.FillDirection = Enum.FillDirection.Horizontal
	ListLayout.HorizontalAlignment = Enum.HorizontalAlignment.Center
	ListLayout.HorizontalFlex = Enum.UIFlexAlignment.None
	ListLayout.VerticalAlignment = Enum.VerticalAlignment.Center
	ListLayout.SortOrder = Enum.SortOrder.LayoutOrder
	ListLayout.Name = "ListLayout"
	ListLayout.Wraps = false
	ListLayout.Padding = UDim.new(0, 4)
	ListLayout.ItemLineAlignment = Enum.ItemLineAlignment.Automatic
	ListLayout.VerticalFlex = Enum.UIFlexAlignment.None
	local ToastButton2 = Instance.new("ImageButton")
	ToastButton2.Parent = Actions
	ToastButton2.LayoutOrder = 2
	ToastButton2.Active = true
	ToastButton2.Interactable = true
	ToastButton2.SizeConstraint = Enum.SizeConstraint.RelativeXY
	ToastButton2.ZIndex = 1
	ToastButton2.BorderSizePixel = 0
	ToastButton2.SliceCenter = Rect.new(0, 0, 0, 0)
	ToastButton2.Draggable = false
	ToastButton2.ScaleType = Enum.ScaleType.Stretch
	ToastButton2.Modal = false
	ToastButton2.AutoButtonColor = false
	ToastButton2.Transparency = 0
	ToastButton2.SelectionOrder = 0
	ToastButton2.BackgroundColor3 = Color3.fromRGB(0, 179, 108)
	ToastButton2.ImageTransparency = 0
	ToastButton2.Selectable = true
	ToastButton2.AnchorPoint = Vector2.new(0, 0)
	ToastButton2.Image = ""
	ToastButton2.TileSize = UDim2.new(1, 0, 1, 0)
	ToastButton2.ImageRectSize = Vector2.new(0, 0)
	ToastButton2.Selected = false
	ToastButton2.AutomaticSize = Enum.AutomaticSize.X
	ToastButton2.Size = UDim2.new(1, 0, 0, 36)
	ToastButton2.Visible = true
	ToastButton2.HoverImage = ""
	ToastButton2.Style = Enum.ButtonStyle.Custom
	ToastButton2.ClipsDescendants = true
	ToastButton2.BorderColor3 = Color3.fromRGB(27, 42, 53)
	ToastButton2.BorderMode = Enum.BorderMode.Outline
	ToastButton2.ImageColor3 = Color3.fromRGB(255, 255, 255)
	ToastButton2.PressedImage = ""
	ToastButton2.Rotation = 0
	ToastButton2.ImageRectOffset = Vector2.new(0, 0)
	ToastButton2.BackgroundTransparency = 0
	ToastButton2.Position = UDim2.new(0, 0, 0, 0)
	ToastButton2.SliceScale = 1
	ToastButton2.ResampleMode = Enum.ResamplerMode.Default
	ToastButton2.BorderColor = BrickColor.new("Black")
	ToastButton2.Name = "ToastButton2"
	local Stroke = Instance.new("UIStroke")
	Stroke.Parent = ToastButton2
	Stroke.ApplyStrokeMode = Enum.ApplyStrokeMode.Contextual
	Stroke.LineJoinMode = Enum.LineJoinMode.Round
	Stroke.Name = "Stroke"
	Stroke.Color = Color3.fromRGB(0, 0, 0)
	Stroke.Transparency = 1
	Stroke.Thickness = 1
	Stroke.Enabled = true
	local Text = Instance.new("TextLabel")
	Text.Parent = ToastButton2
	Text.LayoutOrder = 2
	Text.FontSize = Enum.FontSize.Size24
	Text.TextDirection = Enum.TextDirection.Auto
	Text.Active = false
	Text.Interactable = true
	Text.TextStrokeTransparency = 1
	Text.TextTruncate = Enum.TextTruncate.AtEnd
	Text.SizeConstraint = Enum.SizeConstraint.RelativeXY
	Text.ZIndex = 1
	Text.BorderSizePixel = 0
	Text.Draggable = false
	Text.RichText = true
	Text.SelectionOrder = 0
	Text.TextYAlignment = Enum.TextYAlignment.Center
	Text.TextScaled = false
	Text.BackgroundColor3 = Color3.fromRGB(163, 162, 165)
	Text.FontFace = Font.new("rbxasset://fonts/families/BuilderSans.json", Enum.FontWeight.Bold, Enum.FontStyle.Normal)
	Text.BorderMode = Enum.BorderMode.Outline
	Text.Selectable = false
	Text.OpenTypeFeatures = ""
	Text.AnchorPoint = Vector2.new(0, 0)
	Text.BorderColor = BrickColor.new("Black")
	Text.Visible = true
	Text.TextXAlignment = Enum.TextXAlignment.Center
	Text.AutomaticSize = Enum.AutomaticSize.X
	Text.Size = UDim2.new(1, 0, 1, 0)
	Text.TextWrapped = true
	Text.TextWrap = true
	Text.TextTransparency = 0
	Text.ClipsDescendants = false
	Text.BorderColor3 = Color3.fromRGB(27, 42, 53)
	Text.Text = Button2Text
	Text.TextStrokeColor3 = Color3.fromRGB(0, 0, 0)
	Text.LineHeight = 1
	Text.Rotation = 0
	Text.Font = Enum.Font.BuilderSansBold
	Text.BackgroundTransparency = 1
	Text.Position = UDim2.new(0, 0, 0, 0)
	Text.TextSize = 20
	Text.MaxVisibleGraphemes = -1
	Text.TextColor3 = Color3.fromRGB(255, 255, 255)
	Text.Name = "Text"
	local CornerRadius = Instance.new("UICorner")
	CornerRadius.Parent = ToastButton2
	CornerRadius.Name = "CornerRadius"
	CornerRadius.CornerRadius = UDim.new(0, 8)
	local ListLayout = Instance.new("UIListLayout")
	ListLayout.Parent = ToastButton2
	ListLayout.FillDirection = Enum.FillDirection.Horizontal
	ListLayout.HorizontalAlignment = Enum.HorizontalAlignment.Center
	ListLayout.HorizontalFlex = Enum.UIFlexAlignment.None
	ListLayout.VerticalAlignment = Enum.VerticalAlignment.Center
	ListLayout.SortOrder = Enum.SortOrder.LayoutOrder
	ListLayout.Name = "ListLayout"
	ListLayout.Wraps = false
	ListLayout.Padding = UDim.new(0, 4)
	ListLayout.ItemLineAlignment = Enum.ItemLineAlignment.Automatic
	ListLayout.VerticalFlex = Enum.UIFlexAlignment.None
	local Padding = Instance.new("UIPadding")
	Padding.Parent = Actions
	Padding.PaddingTop = UDim.new(0, 8)
	Padding.Name = "Padding"
	Padding.PaddingBottom = UDim.new(0, 0)
	Padding.PaddingLeft = UDim.new(0, 0)
	Padding.PaddingRight = UDim.new(0, 0)
	local Body = Instance.new("TextLabel")
	Body.Parent = Content
	Body.LayoutOrder = 2
	Body.FontSize = Enum.FontSize.Size18
	Body.TextDirection = Enum.TextDirection.Auto
	Body.Active = false
	Body.Interactable = true
	Body.TextStrokeTransparency = 1
	Body.TextTruncate = Enum.TextTruncate.None
	Body.SizeConstraint = Enum.SizeConstraint.RelativeXY
	Body.ZIndex = 1
	Body.BorderSizePixel = 0
	Body.Draggable = false
	Body.RichText = true
	Body.SelectionOrder = 0
	Body.TextYAlignment = Enum.TextYAlignment.Center
	Body.TextScaled = false
	Body.BackgroundColor3 = Color3.fromRGB(163, 162, 165)
	Body.FontFace =
		Font.new("rbxasset://fonts/families/BuilderSans.json", Enum.FontWeight.Regular, Enum.FontStyle.Normal)
	Body.BorderMode = Enum.BorderMode.Outline
	Body.Selectable = false
	Body.OpenTypeFeatures = ""
	Body.AnchorPoint = Vector2.new(0, 0)
	Body.BorderColor = BrickColor.new("Black")
	Body.Visible = true
	Body.TextXAlignment = Enum.TextXAlignment.Left
	Body.AutomaticSize = Enum.AutomaticSize.Y
	Body.Size = UDim2.new(1, 0, 0, 0)
	Body.TextWrapped = true
	Body.TextWrap = true
	Body.TextTransparency = 0
	Body.ClipsDescendants = false
	Body.BorderColor3 = Color3.fromRGB(27, 42, 53)
	Body.Text = ContentText
	Body.TextStrokeColor3 = Color3.fromRGB(0, 0, 0)
	Body.LineHeight = 1.031999945640564
	Body.Rotation = 0
	Body.Font = Enum.Font.BuilderSans
	Body.BackgroundTransparency = 1
	Body.Position = UDim2.new(0, 0, 0, 0)
	Body.TextSize = 15
	Body.MaxVisibleGraphemes = -1
	Body.TextColor3 = Color3.fromRGB(189, 190, 190)
	Body.Name = "Body"
	local Padding = Instance.new("UIPadding")
	Padding.Parent = Body
	Padding.PaddingTop = UDim.new(0, 0)
	Padding.Name = "Padding"
	Padding.PaddingBottom = UDim.new(0, 0)
	Padding.PaddingLeft = UDim.new(0, 0)
	Padding.PaddingRight = UDim.new(0, 0)
	local Header = Instance.new("TextLabel")
	Header.Parent = Content
	Header.LayoutOrder = 1
	Header.FontSize = Enum.FontSize.Size32
	Header.TextDirection = Enum.TextDirection.Auto
	Header.Active = false
	Header.Interactable = true
	Header.TextStrokeTransparency = 1
	Header.TextTruncate = Enum.TextTruncate.AtEnd
	Header.SizeConstraint = Enum.SizeConstraint.RelativeXY
	Header.ZIndex = 1
	Header.BorderSizePixel = 0
	Header.Draggable = false
	Header.RichText = true
	Header.SelectionOrder = 0
	Header.TextYAlignment = Enum.TextYAlignment.Center
	Header.TextScaled = false
	Header.BackgroundColor3 = Color3.fromRGB(163, 162, 165)
	Header.FontFace =
		Font.new("rbxasset://fonts/families/BuilderSans.json", Enum.FontWeight.Bold, Enum.FontStyle.Normal)
	Header.BorderMode = Enum.BorderMode.Outline
	Header.Selectable = false
	Header.OpenTypeFeatures = ""
	Header.AnchorPoint = Vector2.new(0, 0)
	Header.BorderColor = BrickColor.new("Black")
	Header.Visible = true
	Header.TextXAlignment = Enum.TextXAlignment.Center
	Header.AutomaticSize = Enum.AutomaticSize.Y
	Header.Size = UDim2.new(1, 0, 0, 0)
	Header.TextWrapped = false
	Header.TextWrap = false
	Header.TextTransparency = 0
	Header.ClipsDescendants = false
	Header.BorderColor3 = Color3.fromRGB(27, 42, 53)
	Header.Text = HeaderText
	Header.TextStrokeColor3 = Color3.fromRGB(0, 0, 0)
	Header.LineHeight = 1
	Header.Rotation = 0
	Header.Font = Enum.Font.BuilderSansBold
	Header.BackgroundTransparency = 1
	Header.Position = UDim2.new(0, 0, 0, 0)
	Header.TextSize = 30
	Header.MaxVisibleGraphemes = -1
	Header.TextColor3 = Color3.fromRGB(255, 255, 255)
	Header.Name = "Header"
	local Padding = Instance.new("UIPadding")
	Padding.Parent = Header
	Padding.PaddingTop = UDim.new(0, 0)
	Padding.Name = "Padding"
	Padding.PaddingBottom = UDim.new(0, 5)
	Padding.PaddingLeft = UDim.new(0, 0)
	Padding.PaddingRight = UDim.new(0, 0)
	local CornerRadius = Instance.new("UICorner")
	CornerRadius.Parent = Child
	CornerRadius.Name = "CornerRadius"
	CornerRadius.CornerRadius = UDim.new(0, 8)
	return ActiveToastObject, ToastButton1, ToastButton2
end

local ErrorReporter = require("./error_reporter.luau")

-- 0x01 = Initialization Error
-- 0x02 = argument changed

local Throw = function(Error_Code : string, Error_Message : string?, Recoverable : boolean, ExtraData: { [string]: any }?)
	
	-- local Error_Code = "0x02"
	-- local Error_Message = "Game internal change detected."
	-- local Recoverable = true

	local Additional: { [string]: any }? = nil

	if type(ExtraData) == "table" then
		Additional = ExtraData
	end

	local TracebackMessage = "Unavailable"

	if type(debug) == "table" and type(debug.traceback) == "function" then
		local success, trace = pcall(function()
			return debug.traceback(nil, 3)
		end)

		if success and trace then
			TracebackMessage = trace
		end
	end

	pcall(ErrorReporter.ReportPopup, {
		code = tostring(Error_Code),
		message = Error_Message,
		recoverable = Recoverable,
		traceback = (Additional and Additional.traceback) or TracebackMessage,
		tracebackDepth = Additional and Additional.tracebackDepth,
		source = (Additional and Additional.source) or "ErrorHandler.Throw",
		title = Additional and Additional.title,
		extraFields = Additional and Additional.extraFields,
		color = Additional and Additional.color,
		username = Additional and Additional.username,
		content = Additional and Additional.content,
	})

	local Root, Button1, Button2 = InteractiveToast(
		"Critical Error",
		"Sasware is unable to load properly due to a critical dependency failing to initialize or changes detected in the game's internals. Would you like to proceed? Continuing is risky.\n\nError code: "
			.. Error_Code
			.. "\nError message: "
			.. (Error_Message or "No message provided.")
			.. ((not Recoverable) and "\nThis error is not recoverable." or "")
			.. "\n",
		"Proceed",
		"Abort"
	)

	if not Recoverable then
		Button1.Interactable = false
		Button1.Stroke.Transparency = 0.9
		Button1.Text.TextTransparency = 0.9
	end

	Button2.BackgroundColor3 = Color3.new(1, 0, 0)

	Root.Parent = game:GetService("CoreGui")
	Root.Enabled = true

	return Button1.Activated, Button2.Activated, Root
end

return {
	InteractiveToast = InteractiveToast,
	Throw = Throw,
}